### DISCUSSION TITLE
Which point in space is the Jacobian matrix computed at?

### DISCUSSION THREAD

**Discussion by Wayne-Z** [stack-of-tasks/pinocchio#1975] [Q&A] [answered]
Hello,
    I tried to calculate the Jacobian matrix of the end effector of the kinova-gen3 robot in the world frame to calculate `v_ee` by `v_ee=Je*q_d`, I used `getFrameJacobian(model,data,end_effector_joint_id, WORLD.Je)`  and `getJointJacobian(model,data,7,WORLD,Je)` to get the Jacobian matrix, their results are the same, but different from the correct Jocabian matrix of the end effector, as shown in the figure.

```
Correct Jacobian matrix of EE wrt World frame: v_EE=J*q_d
   0.311402   -0.129344     0.16731   0.0141984   0.0663689  -0.0996134 2.08167e-17
   0.238223   -0.203283     0.26236   -0.241407   0.0482815    0.134534 -3.1225e-17
          0    0.134845   -0.173676    -0.28996   -0.141173  -0.0008199           0
          0    0.843695   -0.261631   -0.632851     0.73305    0.406432    0.693336
          0   -0.536822   -0.411191   -0.609998   -0.670318    0.295668    0.516433
         -1           0   -0.873196    0.476867    0.115374   -0.864519    0.502576
```

```
Jacobian matrix of EndEffector_joint by pinocchio
 -1.1492e-06     0.152892     0.111602     0.483362     0.454702    -0.524243     -0.11523
  -4.322e-19     0.240293   -0.0831721    -0.460524     0.461164      0.14222     0.484543
           0  7.95716e-07   0.00572665    0.0523776    -0.209705    -0.197815    -0.338931
 -2.7629e-18     0.843695    -0.261631    -0.632851     0.733051     0.406426     0.693336
 7.34641e-06    -0.536822    -0.411184    -0.609999    -0.670318     0.295675      0.51643
          -1 -2.70512e-07    -0.873199     0.476866     0.115368    -0.864519     0.502579
```

Then I calculate the jacobian matrix of another link  use `getFrameJacobian`, and I found that the Jacobian matrix of this link is same to the end_effector_joint with the last few columns of the matrix become 0. But they are different point in space and the matrix should be different.

```
Jacobian matrix of HalfArm1_Link
 -1.1492e-06     0.152892            0            0            0            0            0
  -4.322e-19     0.240293            0            0            0            0            0
           0  7.95716e-07            0            0            0            0            0
 -2.7629e-18     0.843695            0            0            0            0            0
 7.34641e-06    -0.536822            0            0            0            0            0
          -1 -2.70512e-07            0            0            0            0            0
```

So, I would like to ask which point in space the Jacobi matrix obtained from Pinocchio corresponds to？Please correct me if I have used it incorrectly.
![2023-05-27 00-10-32屏幕截图](https://github.com/stack-of-tasks/pinocchio/assets/46589349/a31fa83c-7145-483e-8060-9be731360506)

**Comment by Wayne-Z** [comment]
I think I found the issue. The jacobian matrix  in world frame get from pinocchio is relative to `World Coordinate Origin` rather than `target point`.  It means that, it is `$ z_i \times ^i p_o^o $` rather than `$ z_i \times ^i p_n^o $` in the first three rows of J.


![01](https://github.com/stack-of-tasks/pinocchio/assets/46589349/19942e38-f98a-4186-979b-efc306a88f55)
![02](https://github.com/stack-of-tasks/pinocchio/assets/46589349/a37047a4-181e-4ccd-a79c-ab022313726d)

**Comment by jcarpent** [comment, marked as answer]
Yes, you're right. You should work with LOCAL or LOCAL_WORLD_ALIGNED type of frame.

**Reply by Wayne-Z** [reply]
Dear jcarpent,
    Thanks for your advice. I change the frame to LOCAL frame of EndEffector_Link, it is the jacobian of  `EndEffector_Link` w.r.t LOCAL frame!
    I have one more question, if I want get the jacobian of a Fixed_joint, which function should I use: `getJointJacobian()` or `getframeJacobian()`? Pinocchio seems to provide no joint_ID for fixed_joint, only joint_ID for movable joints. 
   Thanks a lot for any help on this.

**Reply by jcarpent** [reply]
Sorry for the late reply. You should indeed use `getFrameJacobian()` for fixed joints represented as frames.
You can also use this function for movable joints (as all the joints have an associated frame in the vector of frames).

**Reply by xiaosuyang** [reply]
For a biped robot ,where is the WORLD frame? Is it moving with baselink(torso)?,
